lib/lf/refinement/refinement_hierarchy.h

```diff
@@ -0,0 +1,249 @@
+   */
+  std::shared_ptr<const mesh::Mesh> getMesh(size_type level) const {
+    LF_VERIFY_MSG(level < meshes_.size(),
+                  "Level " << level << " outside scope");
+    return meshes_.at(level);
+  }
+  std::shared_ptr<mesh::Mesh> getMesh(size_type level) {
+    LF_VERIFY_MSG(level < meshes_.size(),
+                  "Level " << level << " outside scope");
+    return meshes_.at(level);
+  }
+
+  /**
+   * @brief Obtain refinement information for all points
+   */
+  const std::vector<PointChildInfo> &point_child_info(size_type level) const {
+    LF_VERIFY_MSG(level < NumLevels(), "Illegal level " << level);
+    return point_child_infos_[level];
+  }
+  /**
+   * @brief Obtain refinement information for all edges
+   */
+  const std::vector<EdgeChildInfo> &edge_child_info(size_type level) const {
+    LF_VERIFY_MSG(level < NumLevels(), "Illegal level " << level);
+    return edge_child_infos_[level];
+  }
+  /**
+   e* @brief Obtain refinement information for all
+   */
+  const std::vector<CellChildInfo> &cell_child_info(size_type level) const {
+    LF_VERIFY_MSG(level < NumLevels(), "Illegal level " << level);
+    return cell_child_infos_[level];
+  }
+  /**
+   * @brief Fetch information about parents
+   */
+  const std::vector<ParentInfo> &parent_info(size_type level,
+                                             dim_t codim) const {
+    LF_VERIFY_MSG(level < NumLevels(), "Illegal level " << level);
+    LF_VERIFY_MSG(codim < 3, "Codim = " << codim << " illegal");
+    return parent_infos_[level][codim];
+  }
+  /**
+   * @brief Access refinement edge indices
+   */
+  const std::vector<glb_idx_t> &refinement_edges(size_type level) const {
+    LF_VERIFY_MSG(level < NumLevels(), "Illegal level " << level);
+    return refinement_edges_[level];
+  }
+
+  /**
+   * @brief Perform regular or barycentric uniform refinement of the finest mesh
+   * in the hierarchy
+   *
+   * @param ref_pat selector for type of uniform refinement: default is
+   * rp_regular, rp_barycentric choses barycentric refinement.
+   *
+   * This method carries out uniform refinement of all cells of a mesh according
+   * to the `rp_regular` or `rp_barycentric` refinement patterns.
+   *
+   * Regular refinement means that every node is copied, every edge is split
+   * and every cell is subdivided into four or six smaller ones of the same
+   * shape.
+   */
+  void RefineRegular(RefPat ref_pat = RefPat::rp_regular);
+  /**
+   * @brief Mark the edges of a mesh based on a predicate
+   *
+   * @param marker this should be functor of type
+   * `std::function<bool(const Mesh &,const Entity &)>`
+   * returning true if the passed edge is to be marked.
+   *
+   * The _marker_ object also takes a reference to a mesh, because
+   * marking makes sense only for the finest level. The mesh on the
+   * finest level is provided to the marker object by the `MeshHierarchy`.
+   */
+  template <typename Marker>
+  void MarkEdges(Marker &&marker);
```

Comment:
I prefer to keep the current signature, because marking makes sense only for the finest mesh in a hierarchy and this finest mesh is picked by the MarkEdges() method. Passing a marker object with an inconsistent mesh could be a source of errors hard to track.

Reply:
I see, this is of course a problem, but I think this problem is not fully resolved by passing the grid along as well. This resolves the problem only in some special cases, but other cases can still happen: For example, the user may construct a `MeshDataSet<bool>` to mark the edges but construct it on the wrong mesh:
```
MeshHierarchy mh(...);
std::shared_ptr<MeshDataSet<bool>> mds = make_CodimMeshDataSet(mh.getMesh(0));
mh.RefineRegular();
mh.markEdges([&](const auto&, const auto&) {return (*mds)(e);}); // error here
```

How about introducing a `LF_ASSERT_MSG(Contains(e), "entity not contained in this mesh")` to the `lf::mesh::hybrid2dp::Mesh::Index()` method?

Also an option would be to inverse control and replace `MarkEdges()` with the method
```
void MarkEdge(const Entity& edge)
```
which would mark that specific edge for refinement...? This method could of course also check whether the passed edge belong to the right mesh via the `lf::mesh::Mesh::Contains()` method.